external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala

```diff
@@ -73,12 +73,14 @@ class DirectKafkaInputDStream[
 
   protected val kc = new KafkaCluster(kafkaParams)
 
-  protected val maxMessagesPerPartition: Option[Long] = {
-    val ratePerSec = context.sparkContext.getConf.getInt(
+  private val ratePerSec: Int = context.sparkContext.getConf.getInt(
       "spark.streaming.kafka.maxRatePerPartition", 0)
-    if (ratePerSec > 0) {
+  protected def maxMessagesPerPartition: Option[Long] = {
+    val estimatedRate = rateController.getLatestRate().toInt
+    val effectiveRatePerSec = if (estimatedRate > 0) (ratePerSec min estimatedRate) else ratePerSec
```

Comment:
The estimated rate is per stream, not per Kafka partition (like the configuration parameter above). Shouldn't this be accounted for in here?